task/control-plane-agent/src/dump.rs

```diff
@@ -0,0 +1,276 @@
+        else {
+            return Err(SpError::Dump(DumpError::BadKey));
+        };
+
+        // Figure out if we're replaying a previous chunk or not
+        let replay = if seq == state.expected_seq {
+            false
+        } else if state.expected_seq > 0 && seq == state.expected_seq - 1 {
+            true
+        } else {
+            return Err(SpError::Dump(DumpError::BadSequenceNumber));
+        };
+
+        // We'll update this position as we go, then write it back afterwards
+        let mut pos = if replay {
+            state.prev_pos.unwrap_lite()
+        } else {
+            state.current_pos
+        };
+
+        let mut header = humpty::DumpAreaHeader::new_zeroed();
+        self.agent
+            .read_dump_into(pos.area_index, 0, header.as_bytes_mut())
+            .map_err(|_e| SpError::Dump(DumpError::ReadFailed))?;
+
+        // Make sure the header is still valid
+        if header.contents != humpty::DumpContents::SingleTask.into() {
+            return Err(SpError::Dump(DumpError::NoLongerValid));
+        }
+
+        // Move along to the next area if we're at the end
+        if pos.offset + SEGMENT_DATA_SIZE > header.written {
+            if header.next == 0 {
+                // we're done, because there's no more dump areas
+                return Ok(None);
+            }
+
+            // Move to the next area and read the header
+            pos.area_index += 1;
+            self.agent
+                .read_dump_into(pos.area_index, 0, header.as_bytes_mut())
+                .map_err(|_e| SpError::Dump(DumpError::ReadFailed))?;
+
+            // If the next header is of a different type, then we have no more
+            // data left to read and can return None.
+            if header.contents != humpty::DumpContents::SingleTask.into()
+                || header.nsegments != 0
+            {
+                return Ok(None);
+            }
+            // Skip the area header
+            pos.offset = HEADER_SIZE;
+        }
+
+        // Read the dump segment data header
+        let mut ds = humpty::DumpSegmentData::new_zeroed();
+        self.agent
+            .read_dump_into(pos.area_index, pos.offset, ds.as_bytes_mut())
+            .map_err(|_e| SpError::Dump(DumpError::ReadFailed))?;
+        pos.offset += SEGMENT_DATA_SIZE;
+
+        // Read the compressed bytes directly into the tx buffer
+        if ds.compressed_length as usize > buf.len() {
+            return Err(SpError::Dump(DumpError::SegmentTooLong));
+        }
+        self.agent
+            .read_dump_into(
+                pos.area_index,
+                pos.offset,
+                &mut buf[..ds.compressed_length as usize],
+            )
+            .map_err(|_e| SpError::Dump(DumpError::ReadFailed))?;
+        pos.offset += ds.compressed_length as u32;
+        while pos.offset & 3 != 0 {
+            // pad to the nearest u32
+            // XXX why is `humpty::DUMP_SEGMENT_MASK` private?
```

Comment:
I think the answer there was purely historical (this happened to be the first constant in that file); will push a fix to that to `humpty` and then you can use the constant here.

Reply:
Bumped and using the constant now